Implement CSS text-transform: full-width

mdubet opened the pull request:
#### 2b128e4b33892acb61b70cf325cfdbbc14717dab
<pre>
Implement CSS text-transform: full-width
<a href="https://bugs.webkit.org/show_bug.cgi?id=143153">https://bugs.webkit.org/show_bug.cgi?id=143153</a>
rdar://81520147

Reviewed by Myles C. Maxfield.

Only on Cocoa platform. Some characters mapping are missing from the underlying API, such as &apos;(&apos; and &apos;)&apos;.

<a href="https://w3c.github.io/csswg-drafts/css-text/#full-width">https://w3c.github.io/csswg-drafts/css-text/#full-width</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-transform-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/platform/graphics/TextTransform.cpp: Added.
(WebCore::transformToFullWidth):
* Source/WebCore/platform/graphics/TextTransform.h: Added.
* Source/WebCore/platform/graphics/cocoa/TextTransformCocoa.cpp: Added.
(WebCore::transformToFullWidth):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: <a href="https://commits.webkit.org/261211@main">https://commits.webkit.org/261211@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c8592a222742f1f2bb3d7b6f114111ec675fdcd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19884 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43346 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2148 "Hash 8c8592a2 for PR 10862 does not build (failure)") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119616 "Hash 8c8592a2 for PR 10862 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114763 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10989 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/87/builds/2148 "Hash 8c8592a2 for PR 10862 does not build (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15862 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99020 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103156 "Hash 8c8592a2 for PR 10862 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97822 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30686 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/103156 "Hash 8c8592a2 for PR 10862 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12476 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32022 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/103156 "Hash 8c8592a2 for PR 10862 does not build (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13045 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9001 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18428 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7777 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14999 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -5994,6 +5994,7 @@
                 "uppercase",
                 "lowercase",
                 "full-size-kana",
+                "full-width",
```
```suggestion
                {
                    "value": "full-width",
                    "enable-if": "PLATFORM(COCOA)"
                },
```

Maybe do something like this to allow feature detection to properly work.

pxlcoder commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 54;
+	objectVersion = 51;
```
I don't think we should be making changes to `objectVersion`.

cc @emw-apple

emw-apple replied:
`54` is the archiver version for Big Sur-aligned Xcode, which is the oldest Apple platform we support, so it should be safe to use here.

mdubet replied:
I'm gonna remove it anyway, it should not be in this PR.

litherum commented on LayoutTests/TestExpectations:
```diff
@@ -2774,13 +2774,14 @@ imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-eleme
-webkit.org/b/143153 imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullwidth-008.html [ ImageOnlyFailure ]
-webkit.org/b/143153 imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullwidth-009.html [ ImageOnlyFailure ]
+# text-transform: full-width
+# no full width mapping for characters ( and )
+imported/w3c/web-platform-tests/css/css-text/text-transform/text-transform-fullwidth-001.xht [ ImageOnlyFailure ]
+# full width and space concatenation issue
```
Why no bug links?

litherum commented on Source/WebCore/platform/graphics/TextTransformation.h:
```diff
@@ -0,0 +1,49 @@
+inline String transformToFullWidth(const String& text)
+{
+#if PLATFORM(COCOA)
+    CFMutableStringRef mutableString = CFStringCreateMutableCopy(nullptr, 0, text.createCFString().get());
```
This is a memory leak. You need `adoptCF()`

litherum commented on Source/WebCore/platform/graphics/TextTransformation.h:
```diff
@@ -0,0 +1,49 @@
+{
+#if PLATFORM(COCOA)
+    CFMutableStringRef mutableString = CFStringCreateMutableCopy(nullptr, 0, text.createCFString().get());
+    if (CFStringTransform(mutableString, nullptr, kCFStringTransformFullwidthHalfwidth, true))
```
Hey that's pretty cool!

litherum commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -5994,6 +5994,10 @@
                 "full-size-kana",
+                {
+                    "value": "full-width",
+                    "enable-if": "PLATFORM(COCOA)"
```
Does this actually make `@supports (text-transform: full-size-kana)` actually false on non-Cocoa platforms?

We should make sure we do what is necessary to make that false on non-Cocoa platforms.

nt1m replied:
That's the intent, although @weinig might know better if this actually works or not.

mdubet replied:
The `PLATFORM(COCOA)` macro doesn't seem to be defined in the parser context. Maybe we should open another PR for this.

litherum commented on Source/WebCore/platform/graphics/TextTransformation.h:
```diff
@@ -0,0 +1,49 @@
+// https://w3c.github.io/csswg-drafts/css-text/#full-width
+inline String transformToFullWidth(const String& text)
+{
+#if PLATFORM(COCOA)
```
Can we put this inside TextTransformCocoa.cpp (and add a TextTransform.h) so we don't need the `#if` guards? We usually use this kind of design for platform-specific code. Other ports can add their own TextTransformFoo.cpp files when they want to.

mdubet replied:
Fixed

mdubet committed:
Implement CSS text-transform: full-width
https://bugs.webkit.org/show_bug.cgi?id=143153
rdar://81520147

Reviewed by Myles C. Maxfield.

Only on Cocoa platform. Some characters mapping are missing from the underlying API, such as '(' and ')'.

https://w3c.github.io/csswg-drafts/css-text/#full-width

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/text-transform-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/platform/graphics/TextTransform.cpp: Added.
(WebCore::transformToFullWidth):
* Source/WebCore/platform/graphics/TextTransform.h: Added.
* Source/WebCore/platform/graphics/cocoa/TextTransformCocoa.cpp: Added.
(WebCore::transformToFullWidth):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: https://commits.webkit.org/261211@main
